// Copyright Offene Werkstatt WÃ¤denswil
// SPDX-License-Identifier: MIT

syntax = "proto3";

package maco.secrets;

// RPC service for factory provisioning of device secrets.
// This service is used during device setup to provision cryptographic
// keys. Once provisioned, secrets persist across reboots.
service DeviceSecretsService {
  // Check if secrets have been provisioned.
  rpc GetStatus(Empty) returns (StatusResponse);

  // Provision device secrets. Fails if already provisioned.
  // Use Clear() first if re-provisioning is needed.
  rpc Provision(ProvisionRequest) returns (ProvisionResponse);

  // Clear all provisioned secrets.
  // WARNING: This erases cryptographic material permanently.
  rpc Clear(Empty) returns (ProvisionResponse);
}

message Empty {}

message ProvisionRequest {
  // Master secret for ASCON key derivation (gateway communication).
  bytes gateway_master_secret = 1;  // Must be exactly 16 bytes

  // NTAG 424 DNA terminal key (KEY_TERMINAL, NTAG slot 1).
  bytes ntag_terminal_key = 2;  // Must be exactly 16 bytes
}

message ProvisionResponse {
  bool success = 1;
  string error = 2;
}

message StatusResponse {
  bool is_provisioned = 1;
}
